### PR TITLE
Introduce developer tools to replace the util scripts

### LINKED ISSUE #1144: NXDL Source link is wrong for contributed definitions
For example:

https://nexusformat.github.io/definitions/classes/contributed_definitions/NXaberration.html#nxaberration

It links to

https://github.com/nexusformat/definitions/blob/main/base_classes/NXaberration.nxdl.xml

but is should be

https://github.com/nexusformat/definitions/blob/main/contributed_definitions/NXaberration.nxdl.xml

### PR DESCRIPTION
**Pull request by woutdenolf**
Closes #1144 #1131

Currently: apart from the actual `sphinx` building we have some scripts in `utils` which in combination with `make` files and some custom stuff in the github workflows prepare the files for a  `sphinx` build.

In this PR I propose a python based CLI to do all the non-`sphinx` stuff

https://github.com/woutdenolf/definitions/blob/dev_tools/dev_tools/README.md

Benefits:

* the CLI has a help and has a README so you don't have to dig through make files and python code to understand what needs to be executed to build the documentation
* better / explicit directory management
* constants are defined only once
* building is always out-of-source
* rst files are directly generated by python instead of piping the stdout to a file
* all NXDL rst files are generated by one command, before it was one command per NXDL file
* speed-up of the anchor list files (saved only once, see previous point)
* we no longer need to copy the nxdl files and `utils` for an out-of-source build
* uses `pytest` to minimize test code (`unittest` has lots of boilerplate)
* easier to make sure CI and a local repo do the same thing

The python scripts in `utils` directory become obsolete:

- [x] `build_preparation.py`: `python3 -m dev_tools manual --prepare` (prepare sphinx build)
- [x] `nxdl2rst.py`: `python3 -m dev_tools manual --prepare` (prepare sphinx build)
- [x] `dev_nxdl2rst.py`: `python3 -m dev_tools nxclass nxfluo --print` (for development)
- [x] `test_nxdl.py`: `pytest dev_tools` (for testing)
- [x] `test_nxdl2rst.py`: `pytest dev_tools` (for testing)
- [x] `test_suite.py`: `pytest dev_tools` (for testing)
- [x] `nxdl_desc2rst.py`:  `python3 -m dev_tools manual --prepare` (prepare sphinx build)
- [x] `nxdl_summary.py`: `python3 -m dev_tools manual --prepare` (prepare sphinx build)
- [x] `types2rst.py`: `python3 -m dev_tools manual --prepare` (prepare sphinx build)
- [x] `units2rst.py`: `python3 -m dev_tools manual --prepare` (prepare sphinx build)
- [x] ~~`dev_units2rst.py`:~~
- [x] ~~`local_utilities.py`:~~

Not included (could be but not in this PR)
- `create_release_notes.py`
- `dev_create_release_notes.py`
- `update_copyright_date.py`

To show the difference with `make local`:

```bash
make local

python3 -m dev_tools manual --prepare --build-root build2
python3 -m dev_tools impatient --prepare --build-root build2

# Difference in generated rst content
python3 -m dev_tools manual --diff

# Difference in directories
diff -qr build build2
```

### CODE REVIEW

**Comment by woutdenolf** [comment]
If this gets accepted, all custom make files can be removed

```
./manual/source/classes/applications/Makefile
./manual/source/classes/Makefile
./manual/source/classes/contributed_definitions/Makefile
./manual/source/classes/base_classes/Makefile
```

The `manual` folder can go back to a normal sphinx structure (no need anymore for the `source` subdirectory)

```
./manual/Makefile  # reset to normal sphinx
./manual/source/Makefile  # remove
```

This one stays the same

```
./impatient-guide/Makefile
```

And the the top make file can use the developer tools

```
./Makefile
```

**Comment by woutdenolf** [comment]
We also only need one github actions workflow (`ci.yaml`). Let me know how you want to proceed with the yml files and the make files.

**Comment by woutdenolf** [comment]
If you have a better name than `dev_tools` let me know.

**Comment by prjemian** [comment]
> Not included (could be but not in this PR)
> 
>     create_release_notes.py
>     dev_create_release_notes.py
>     update_copyright_date.py

Agree we can drop *create_release*.py code.  GitHub has changed its API and this code not longer runs as-is.  A newer, working, version is [in its own repository](https://github.com/prjemian/condatools/blob/main/create_release_notes.py).

We should keep the `update_copyright_date.py` code.  It is used annually. (#969, #876, #713, #611, #529, #460)

**Comment by prjemian** [comment]
Nice!

```bash
(bluesky_2022_2) prjemian@zap:~/.../NeXus/definitions$ python3 -m dev_tools manual --diff | more
--- /home/prjemian/Documents/projects/NeXus/definitions/build/manual/source/classes/base_classes/NXaperture.rst

+++ /home/prjemian/Documents/projects/NeXus/definitions/base_classes/NXaperture.nxdl.xml

@@ -1,4 +1,4 @@

-.. auto-generated by script ../../../../utils/nxdl2rst.py from the NXDL source NXaperture.nxdl.xml

+.. auto-generated by dev_tools.docs.nxdl from the NXDL source base_classes/NXaperture.nxdl.xml -- DO NOT EDIT

 

 .. index::

     ! NXaperture (base class)
```

**Review by prjemian — COMMENTED** [review]
I like the tools.  A lot!  The optimization of the workflow of nxdl2rst is fabulous!

This command history:

```bash
 2034  python3 -m dev_tools impatient --prepare
 2035  python3 -m dev_tools manual --prepare
 2036  sphinx-build -M latexpdf build/impatient-guide/ build/impatient-guide/build
 2037  cp build/impatient-guide/build/latex/NXImpatient.pdf build/manual/source/_static/NXImpatient.pdf
 2038  sphinx-build -M latexpdf build/manual/source/ build/manual/build
 2039  cp build/manual/build/latex/nexus.pdf build/manual/source/_static/NeXusManual.pdf
 2040  man evince
 2041  evince --help
 2042  evince build/manual/build/latex/nexus.pdf &
 2043  sphinx-build -b html -W build/impatient-guide/ build/impatient-guide/build/html
 2044  sphinx-build -b html -W build/manual/source/ build/manual/build/html
 2045  ls build/manual/build/
 2046  pushd build/manual/build/ && tar cf - html | (cd /tmp && tar xf -) && popd
```

reminds me why I will be strongly tempted to re-introduce a `Makefile` to this work.  The point of the `Makefile` is to help the developer remember and execute these commands by implementing a few simple *targets*.  The README.md suggests the names of these targets in its *Quick reference* section.

**Comment by prjemian** [comment]
> If you have a better name than `dev_tools` let me know.

Name is just fine.

**Comment by prjemian** [comment]
@PeterC-DLS What are your comments?

**Comment by woutdenolf** [comment]
> After this is merged, we should apply an rc tag: `v2022.06rc5` then update `CHANGES.rst` before the final release.

Well if this is accepted we should:

- remove all `Makefile` except for the top-level one
- use the `dev_tools` commands in the top-level `Makefile`
- remove to old workflow files
- have `ci.yaml` use `make` instead of the `dev_tools` directly
- remove the obsolete files from `utils`

**Comment by woutdenolf** [comment]
> ```shell
> sphinx-build -M latexpdf build/manual/source/ build/manual/build
> ```
> 
> This is much appreciated, to build PDF with one line and no errors due to nested levels too deep. What made this process correctly?

I didn't dig into what `-M latexpdf` does. All I know is it's the official way of building a PDF (not just the latex files) with `sphinx` and works fine with `sphinx>=5`.

**Comment by woutdenolf** [comment]
Ok as it seemed this proposal is well received, I made to full transition: refactor the make files to use `dev_tools` and remove obsolete files.

**Comment by woutdenolf** [comment]
Note that developers can still use `make` as if nothing changed

```
NeXus: Testing the NXDL files and building the documentation:

make install            Install all requirements to run tests and builds.
make style              Check python coding style.
make autoformat         Format all files to the coding style conventions.
make test               Run NXDL syntax and documentation tests.
make clean              Remove all build files.
make prepare            (Re)create all build files.
make html               Build HTML version of manual. Requires prepare first.
make pdf                Build PDF version of manual. Requires prepare first.
make impatient-guide    Build html & PDF versions of the Guide for the Impatient. Requires prepare first.
make all                Builds complete web site for the manual (in build directory).
make local              (Developer use) Test, prepare and build the HTML manual.

Note:  All builds of the manual will occur in the 'build/' directory.
   For a complete build, run 'make all' in the root directory.
   Developers of the NeXus class definitions can use 'make local' to
   confirm the documentation builds.
```

**Comment by woutdenolf** [comment]
Note that `make style` and `make autoformat` currently only apply to the `dev_tools` directory. In the future we can apply this to the entire repo.

**Comment by PeterC-DLS** [comment]
I checked the html outputs between `main` and this PR and see differences in the generated JSON and YAML vocabulary files:
```
diff -ur build-old/manual/build/html/_static/nxdl_vocabulary.yml build-new/manual/build/html/_static/nxdl_vocabulary.yml
--- build-old/manual/build/html/_static/nxdl_vocabulary.yml     2022-06-28 22:43:31.577263593 +0100
+++ build-new/manual/build/html/_static/nxdl_vocabulary.yml     2022-06-28 22:24:36.550610303 +0100
@@ -1,5 +1,5 @@
 _metadata:
-  datetime: '2022-06-28T21:43:30.934016'
+  datetime: '2022-06-28T21:24:35.906902'
   subtitle: Anchors for all NeXus fields, groups, attributes, and links.
   title: NeXus NXDL vocabulary.
   version: v2022.06
@@ -8,14119 +8,14119 @@
     /NXdata@auxiliary_signals-attribute:
       html: '#nxdata-auxiliary-signals-attribute'
       term: /NXdata@auxiliary_signals-attribute
-      url: https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-auxiliary-signals-attribute
+      url: https://manual.nexusformat.org/classes/base_classes/base_classes/NXdata.html#nxdata-auxiliary-signals-attribute
```
Many of the URLs are different and in JSON, items in the maps are ordered differently. Also SVG files differ, for example:
```
diff -ubr build-old/manual/build/html/_static/412-detector-docs/1d_detector_coordinate_frame.svg build-new/manual/build/html/_static/412-detector-docs/1d_detector_coordinate_frame.svg
--- build-old/manual/build/html/_static/412-detector-docs/1d_detector_coordinate_frame.svg	2022-06-28 22:40:48.099332332 +0100
+++ build-new/manual/build/html/_static/412-detector-docs/1d_detector_coordinate_frame.svg	2022-06-28 22:22:04.838713169 +0100
@@ -16,7 +16,7 @@
    version="1.1"
    inkscape:version="0.48.3.1 r9886"
    sodipodi:docname="1d_detector_coordinate_frame.svg"
-   inkscape:export-filename="/home/wintersb/Development/nexus/definitions/manual/source/img/1d_detector_coordinate_frame.png"
+   inkscape:export-filename="/home/wintersb/Development/nexus/definitions/manual/img/1d_detector_coordinate_frame.png"
    inkscape:export-xdpi="300"
    inkscape:export-ydpi="300">
   <sodipodi:namedview
```

**Comment by woutdenolf** [comment]
> Also SVG files differ, for example

That's something I changed manually.

**Comment by woutdenolf** [comment]
>  differences in the generated JSON, YAML and HTML vocabulary files

Thanks, I'll have a look.

**Comment by woutdenolf** [comment]
> items in the maps are ordered differently

Is that a problem?

**Comment by PeterC-DLS** [comment]
Looks better. Now it seems like the `main` branch generates the incorrect URL:
```
--- build-old/manual/build/html/_downloads/593c638271dc6661be8a6d311d952a7f/nxdl_vocabulary.yml 2022-06-28 22:43:31.577263593 +0100
+++ build/manual/build/html/_downloads/593c638271dc6661be8a6d311d952a7f/nxdl_vocabulary.yml     2022-06-29 19:10:05.839529037 +0100
@@ -1,5 +1,5 @@
 _metadata:
-  datetime: '2022-06-28T21:43:30.934016'
+  datetime: '2022-06-29T18:10:05.195311'
   subtitle: Anchors for all NeXus fields, groups, attributes, and links.
   title: NeXus NXDL vocabulary.
   version: v2022.06
@@ -13,15 +13,15 @@
     /NXcxi_ptycho/DATA@axes-attribute:
       html: '#nxcxi-ptycho-data-axes-attribute'
       term: /NXcxi_ptycho/DATA@axes-attribute
-      url: https://manual.nexusformat.org/classes/applications/NXcxi_ptycho.html#nxcxi-ptycho-data-axes-attribute
+      url: https://manual.nexusformat.org/classes/contributed_definitions/NXcxi_ptycho.html#nxcxi-ptycho-data-axes-attribute
     /NXcxi_ptycho/entry_1/instrument_1/detector_1/translation@axes-attribute:
       html: '#nxcxi-ptycho-entry-1-instrument-1-detector-1-translation-axes-attribute'
       term: /NXcxi_ptycho/entry_1/instrument_1/detector_1/translation@axes-attribute
-      url: https://manual.nexusformat.org/classes/applications/NXcxi_ptycho.html#nxcxi-ptycho-entry-1-instrument-1-detector-1-translation-axes-attribute
+      url: https://manual.nexusformat.org/classes/contributed_definitions/NXcxi_ptycho.html#nxcxi-ptycho-entry-1-instrument-1-detector-1-translation-axes-attribute
```

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+          ls -lAFgh build/impatient-guide/build/latex/NXImpatient.pdf
+
+      - name: Build User Manual
+        run: |
+          make pdf
+          make html
+          ls -lAFgh build/manual/build/html/index.html
+          ls -lAFgh build/manual/build/latex/nexus.pdf
+
+      - name: Build and Commit the User Manual
```
Comment (prjemian) [inline .github/workflows/ci.yaml:83]: consider PR #1147 here

**Comment by prjemian** [comment]
Changes from `manual/source/*` to `manual/*` (such as in changeset ba5625c) are not necessary for this PR.  They make the number of files changed much larger.  IMO, that should happen in a different PR.  Both forms are *standard* sphinx configurations, this choice is an opinion.  Since we now ensure that all builds are out of source, the change eliminates one level of directory clutter.

While I very much like the changes introduced in this PR, review and merge will stall the v2022.06 release which is intended before the end of this month.

We should move this to the next release milestone.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (woutdenolf) [reply .github/workflows/ci.yaml:83]: Mean what? CNAME should be created by the prepare stage?

Comment (woutdenolf) [reply .github/workflows/ci.yaml:83]: Ok I added the generation of the file to the prepare stage of the manual

**Comment by woutdenolf** [comment]
> Changes from `manual/source/*` to `manual/*` (such as in changeset [ba5625c](https://github.com/nexusformat/definitions/commit/ba5625ccaf82e5329a151ec7ec531b6bf6fe0f7e)) are not necessary for this PR. They make the number of files changed much larger. IMO, that should happen in a different PR. Both forms are _standard_ sphinx configurations, this choice is an opinion. Since we now ensure that all builds are out of source, the change eliminates one level of directory clutter.

@prjemian I removed those changes as you requested.

**Comment by woutdenolf** [comment]
If we wait with this until after the release, make sure to merge #1147 and fix #1144.

**Comment by woutdenolf** [comment]
@prjemian I've put the CNAME file in the same directory as #1147

`echo manual.nexusformat.org > ./build/manual/source/CNAME`

But in fact that's not correct is it? It should be

`echo manual.nexusformat.org > ./build/manual/build/html/CNAME`

Right?

**Comment by prjemian** [comment]
Looking at the [advice](https://docs.github.com/en/pages/configuring-a-custom-domain-for-your-github-pages-site/troubleshooting-custom-domains-and-github-pages#cname-errors), I believe you are correct.

> Custom domains are stored in a CNAME file in the root of your publishing source. 

Not the root of the repository but the root of the website to be published.

If writing to the root of the documentation, we must ensure this file is copied to the published site directory.  What @woutdenolf proposes, writing to `./build/manual/build/html/CNAME`, agrees with the advice.

**Comment by woutdenolf** [comment]
I fixed this PR to write the custom domain file to `./build/manual/build/html/CNAME`

**Comment by woutdenolf** [comment]
Alternatively you could use a sphinx extension:

https://www.sphinx-doc.org/en/master/usage/extensions/githubpages.html

@prjemian maybe cleaner?

**Comment by prjemian** [comment]
> Make it so.
>   - Picard

**Comment by woutdenolf** [comment]
> BUT, is it quick to refactor to this

It is done.

**Comment by PeterC-DLS** [comment]
Another niggle now is vocabulary has extra slash after domain:
```
     /NXdata@auxiliary_signals-attribute:
       html: '#nxdata-auxiliary-signals-attribute'
       term: /NXdata@auxiliary_signals-attribute
-      url: https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-auxiliary-signals-attribute
+      url: https://manual.nexusformat.org//classes/base_classes/NXdata.html#nxdata-auxiliary-signals-attribute

```

I prefer to have `main` and this PR produce as similar output as possible before merging but it's just a preference.

**Comment by woutdenolf** [comment]
> Another niggle now is vocabulary has extra slash after domain:

Pull the branch again. I just fixed that.

**Comment by woutdenolf** [comment]
So I set `html_baseurl = "https://manual.nexusformat.org/"` in `conf.py` and it generates the CNAME file. The downside is that this is also used for relative links. Which means if you build the HTML locally and you want all the links resolve locally, that won't work. So maybe the manual CNAME file is better?

**Comment by PeterC-DLS** [comment]
Re JSON vocabulary differences, I wonder if JSON parsers obey map item ordering (Python switched to dictionaries preserving insertion order in key iterations).

**Comment by woutdenolf** [comment]
> So I set `html_baseurl = "https://manual.nexusformat.org/"` in `conf.py` and it generates the CNAME file. The downside is that this is also used for relative links. Which means if you build the HTML locally and you want all the links resolve locally, that won't work. So maybe the manual CNAME file is better?

I reverted back to manual CNAME file creation and removed `html_baseurl` so that links on a local HTML resolve locally.

**Comment by prjemian** [comment]
> Re JSON vocabulary differences, I wonder if JSON parsers obey map item ordering (Python switched to dictionaries preserving insertion order in key iterations).

It is not obvious whether or not order is important.  I only provided both JSON and YAML versions since the intended use was not clear.  Order of items in the JSON file should not be an obstacle now.

**Comment by woutdenolf** [comment]
> Re JSON vocabulary differences, I wonder if JSON parsers obey map item ordering (Python switched to dictionaries preserving insertion order in key iterations).

I did `json.dump(..., sort_keys=True)`. Should be better now.

**Comment by woutdenolf** [comment]
> > Re JSON vocabulary differences, I wonder if JSON parsers obey map item ordering (Python switched to dictionaries preserving insertion order in key iterations).
> 
> I did `json.dump(..., sort_keys=True)`. Should be better now.

Only the terms `tilt_1` and `tilt_2` have moved (and I would say to a better position).

**Comment by prjemian** [comment]
@PeterC-DLS Are we good to merge this today?  Or put it off until next milestone?  If the latter, then merge #1147.  We also need to update `/CHANGES.rst` for the release.

**Comment by woutdenolf** [comment]
> @PeterC-DLS Are we good to merge this today? Or put it off until next milestone? If the latter, then merge #1147. We also need to update `/CHANGES.rst` for the release.

First fix it and the merge. Add also the wrong links (https://github.com/nexusformat/definitions/issues/1144) need to be fixed.

**Comment by prjemian** [comment]
> > @PeterC-DLS Are we good to merge this today? Or put it off until next milestone? If the latter, then merge #1147. We also need to update `/CHANGES.rst` for the release.
> 
> First fix it and the merge. Add also the wrong links (#1144) need to be fixed.

I do not understand what to fix.  Are you referring to #1147

**Comment by PeterC-DLS** [comment]
I think Wout is saying we need to fix the wrong category source URL issue with your suggestion in https://github.com/nexusformat/definitions/pull/1133#issuecomment-1170371459 before the merge then this PR will not change `build/html` output directory significantly. It means this refactor does not fix anything.

**Comment by woutdenolf** [comment]
Sorry I wasn't clear:

1. #1147: it should be `echo manual.nexusformat.org > ./build/manual/build/html/CNAME`. It is still currently `echo manual.nexusformat.org > ./build/manual/source/CNAME`
2. #1144: as described in the issue, some links which should point to `contributed_definitions` point to either `base_classes` or `applications`

Both are fixed in the PR. If we decide to wait with this PR, we need to fix these two issues before making a release.

**Comment by woutdenolf** [comment]
> I think Wout is saying we need to fix the wrong category source URL issue with your suggestion in [#1133 (comment)](https://github.com/nexusformat/definitions/pull/1133#issuecomment-1170371459) before the merge then this PR will not change `build/html` output directory significantly. It means this refactor does not fix anything.

To be clear, this PR fixes all know URL and CNAME issues.

**Comment by prjemian** [comment]
@PeterC-DLS was saying this PR should not fix existing problems but refactor the publishing software from `utils/` to `dev_tools/`, as the title says.

**Comment by woutdenolf** [comment]
> @PeterC-DLS was saying this PR should not fix existing problems

Well I would have to try really hard to preserve the current issues in the `main` branch. The refactoring deal with directories differently so the part that is buggy works differently in the PR. In other words, this fix in the PR cannot simply be used in the `main` branch.

**Comment by prjemian** [comment]
Will this merge today so we can release today?  Are any of these additional changes necessary, or they only to help review?  I believe this PR fixes all existing problems.

**Comment by woutdenolf** [comment]
And here is the other fix: https://github.com/nexusformat/definitions/pull/1151

**Comment by woutdenolf** [comment]
@PeterC-DLS @prjemian 

To summarize:

- All CNAME, URL's and anchor list fixes have been solved in the `main` branch
- I rebased this `dev_tools` branch on the `main` branch
- All CNAME, URL's and anchor list fixes

So this PR is no longer mandatory to make the release. Up to you.

**Comment by woutdenolf** [comment]
It's because I'm rebasing this branch on the main branch whenever we merge something in the main branch. Do this

```
git pull --rebase
```

**Comment by prjemian** [comment]
I'm building locally now for review ...

**Comment by PeterC-DLS** [comment]
Just noticed `listing_category` shouldn't be used in L559 as that makes optional/recommeneded wrong (I think)

**Comment by PeterC-DLS** [comment]
Can check `NXcsg` too.

**Comment by woutdenolf** [comment]
Rebased in `main` branch to include the CNAME fix of https://github.com/nexusformat/definitions/issues/1154

**Comment by woutdenolf** [comment]
Currently the difference with the main branch: [diff.log](https://github.com/nexusformat/definitions/files/9023497/test.log)

```
git checkout main
make local
mv build buildmain

git checkout dev_tools
make local

diff -r buildmain/manual/ build/manual/

**Comment by woutdenolf** [comment]
Without the auto-generated directive: [diff.log](https://github.com/nexusformat/definitions/files/9023539/test.log)

```
diff -r buildmain/manual/ build/manual/ -I 'auto-generated'
```

**Comment by prjemian** [comment]
Captures my attention:

```diff
28c28
<     see: attribute; NXDL attribute
---
>     see:attribute; NXDL attribute
```

Is this deleted whitespace important to Sphinx?

**Review by PeterC-DLS — APPROVED** [review]
Done a comparison with `main` and only differences are timestamps and auto-generated. Good work on the massive speed-up, Wout!

**Comment by prjemian** [comment]
Most of the differences I see are due to trailing white space.  Purely cosmetic.

**Comment by PeterC-DLS** [comment]
`$  diff -ubBr -I 'automatically generated' -I auto-generated -I datetime build-main/manual/build/html/ build/manual/build/html/`

**Comment by prjemian** [comment]
> Done a comparison with `main` and only differences are timestamps and auto-generated. Good work on the massive speed-up, Wout!

By my unscientific measure, about 15x faster!

**Comment by prjemian** [comment]
I pushed the big green button.

**Comment by woutdenolf** [comment]
> Captures my attention:
> 
> ```diff
> 28c28
> <     see: attribute; NXDL attribute
> ---
> >     see:attribute; NXDL attribute
> ```
> 
> Is this deleted whitespace important to Sphinx?

Hmmm I wonder. I didn't catch that.

**Comment by prjemian** [comment]
Locally, built the documentation.  It's OK in the index under "attribute", among other items, says *see NXDL attribute*.

**Comment by PeterC-DLS** [comment]
I don't think it matters: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-index

**Comment by PeterC-DLS** [comment]
Good to release then with 208 items closed in the milestone!

**Comment by prjemian** [comment]
That was a great collaboration, y'all.  The end product was much stronger for all our contributions.

Not to rain on anyone's parade but we have one necessary step before making the release:

- #1165

**Pull request merged**